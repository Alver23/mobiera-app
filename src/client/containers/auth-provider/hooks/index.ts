// Dependencies
import React from 'react';

// Commons
import {
  clearAuthData,
  getAuthData,
} from '@mobiera/containers/login-form/commons';

// Interfaces
import { IUser } from '@mobiera/services/authentication/interfaces';
import IMutableRefObject from '@mobiera/interfaces/mutable-ref';
import IAuthContext from '../context/interfaces';

// Contexts
import AuthContext from '../context';

const useAuth = (): [IAuthContext, () => void, (values: any) => void] => {
  const initialState: IAuthContext = React.useMemo(
    () => ({
      authenticated: false,
      initialize: false,
    }),
    []
  );

  const [data, setData] = React.useState<IAuthContext>(initialState);
  const isMounted: IMutableRefObject<boolean> = React.useRef(true);

  React.useEffect(() => {
    return (): void => {
      isMounted.current = false;
    };
  }, []);

  const logout = React.useCallback((): void => {
    setData({ ...initialState, initialize: true });
    clearAuthData().then();
  }, [initialState]);

  const updateData = React.useCallback((newValues): void => {
    setData((prevState) => ({
      ...prevState,
      ...newValues,
    }));
  }, []);

  React.useEffect((): void => {
    const setAuthData = (user: IUser) => {
      if (isMounted.current) {
        setData({
          user,
          initialize: true,
          authenticated: !!user,
        });
      }
    };

    const setError = () => {
      if (isMounted.current) {
        setData(initialState);
      }
    };

    getAuthData().then(setAuthData).catch(setError);
  }, [initialState]);

  return [data, logout, updateData];
};

export const useAuthSession = (): IAuthContext => {
  return React.useContext<IAuthContext>(AuthContext);
};

export default useAuth;
