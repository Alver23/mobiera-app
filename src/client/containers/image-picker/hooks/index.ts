// Dependencies
import React from 'react';

const useImagePicker = (): [any, () => void] => {
  const actionSheetRef: any = React.useRef();
  const openImagePicker = React.useCallback(() => {
    actionSheetRef.current.show();
  }, []);

  return [actionSheetRef, openImagePicker];
};

export default useImagePicker;
