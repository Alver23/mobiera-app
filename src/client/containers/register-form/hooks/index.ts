// Dependencies
import React from 'react';

// Hooks
import useShowAppLoader from '@mobiera/hooks/app-loader';

// Services
import UserServices from '@mobiera/services/user';

// Interfaces
import { IFormData } from '@mobiera/containers/user-form/interfaces';
import { IFormData as IPayload } from '../interfaces';

const useSignUp = (): [string | undefined, (payload: IFormData) => void] => {
  const [showAppLoader] = useShowAppLoader();
  const [message, setMessage] = React.useState<string>();

  const onSubmit = React.useCallback(
    (payload: IFormData) => {
      const onSuccess = () => {
        setMessage('The user was created successfully');
        showAppLoader(false);
      };

      const onError = () => {
        setMessage('An error has occurred, please try again later');
        showAppLoader(false);
      };

      showAppLoader(true);
      UserServices.save(payload as IPayload)
        .then(onSuccess)
        .catch(onError);
    },
    [showAppLoader]
  );

  return [message, onSubmit];
};

export default useSignUp;
