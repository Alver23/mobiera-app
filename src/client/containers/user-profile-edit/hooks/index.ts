// Dependencies
import React from 'react';

// Hooks
import useShowAppLoader from '@mobiera/hooks/app-loader';

// Services
import UserServices from '@mobiera/services/user';

// Interfaces
import { IFormData } from '@mobiera/containers/user-form/interfaces';
import { IUserUpdateRequest } from '@mobiera/services/user/interfaces';

const useUserUpdate = (
  id: number,
  callback: () => void,
  updateData: (values: any) => void
): [string | undefined, (payload: IFormData) => void] => {
  const [showAppLoader] = useShowAppLoader();
  const [message, setMessage] = React.useState<string>();

  const onSubmit = React.useCallback(
    (payload: IFormData) => {
      const onSuccess = () => {
        setMessage('The user was updated successfully');
        showAppLoader(false);
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { password, ...userPayload } = payload;
        updateData({ user: userPayload });
        callback();
      };

      const onError = () => {
        setMessage('An error has occurred, please try again later');
        showAppLoader(false);
      };

      showAppLoader(true);
      UserServices.update(id, payload as IUserUpdateRequest)
        .then(onSuccess)
        .catch(onError);
    },
    [id, showAppLoader, callback, updateData]
  );

  return [message, onSubmit];
};

export default useUserUpdate;
