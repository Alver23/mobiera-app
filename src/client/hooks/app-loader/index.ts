// Dependencies
import { useCallback } from 'react';
import { useDispatch } from 'react-redux';

// Redux Actions
import appShowLoaderAction from '@mobiera/store/actions/globals';

const useShowAppLoader = (): [(show: boolean) => void] => {
  const dispatch = useDispatch();
  const callback = useCallback(
    (show: boolean) => {
      return dispatch(appShowLoaderAction(show));
    },
    [dispatch]
  );

  return [callback];
};

export default useShowAppLoader;
