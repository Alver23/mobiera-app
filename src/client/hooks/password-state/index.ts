// Dependencies
import React from 'react';

const usePasswordState = (
  initialState = true
): [string, boolean, () => void] => {
  const [showPassword, setShowPassword] = React.useState<boolean>(initialState);
  const iconsMapper: any = {
    true: 'eye',
    false: 'eye-slash',
  };
  const icon = iconsMapper[`${showPassword}`];

  const onChangeSetPassword = React.useCallback(() => {
    setShowPassword(!showPassword);
  }, [showPassword]);

  return [icon, showPassword, onChangeSetPassword];
};

export default usePasswordState;
