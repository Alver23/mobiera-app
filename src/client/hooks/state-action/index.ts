// Dependencies
import React from 'react';

const useStateAction = (
  initialState?: boolean
): [boolean, () => void, () => void] => {
  const [state, setState] = React.useState(!!initialState);
  const truthyCallback = React.useCallback((): void => setState(true), []);
  const falsyCallback = React.useCallback((): void => setState(false), []);
  return [state, truthyCallback, falsyCallback];
};

export default useStateAction;
