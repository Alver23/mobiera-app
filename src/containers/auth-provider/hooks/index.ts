// Dependencies
import React from 'react';

// Interfaces
import IMutableRefObject from '@mobiera/interfaces/mutable-ref';
import IAuthContext from '@mobiera/containers/auth-provider/context/interfaces';

const useAuth = (): [IAuthContext, () => void] => {
  const initialState: IAuthContext = React.useMemo(
    () => ({
      authenticated: false,
      initialize: false,
    }),
    []
  );

  const [data, setData] = React.useState<IAuthContext>(initialState);
  const isMounted: IMutableRefObject<boolean> = React.useRef(true);

  React.useEffect(() => {
    return (): void => {
      isMounted.current = false;
    };
  }, []);

  const logout: () => void = React.useCallback((): void => {
    setData({ ...initialState, initialize: true });
  }, [initialState]);

  React.useEffect((): void => {
    setData((prevstate) => ({
      ...prevstate,
      initialize: true,
    }));
  }, []);

  return [data, logout];
};

export default useAuth;
