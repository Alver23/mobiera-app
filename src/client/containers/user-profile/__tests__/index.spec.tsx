// Dependencies
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Linking } from 'react-native';

// Hooks
import * as AuthHooks from '@mobiera/containers/auth-provider/hooks';

// Interfaces
import IAuthContext from '@mobiera/containers/auth-provider/context/interfaces';

// Components
import UserAvatar from '../components/avatar';
import UserEmail from '../components/email';

// Under test file
import UserProfile from '../index';

// Mocks
import mocks from './mocks.json';

describe('<UserProfile />', () => {
  let component: ShallowWrapper;

  const callback = jest.fn().mockRejectedValue('fake error');
  const sessionMock: IAuthContext = {
    user: mocks,
    authenticated: false,
    initialize: false,
    logout: jest.fn(),
  };

  beforeEach(() => {
    Linking.openURL = callback;
    jest.spyOn(AuthHooks, 'useAuthSession').mockReturnValue(sessionMock);
    component = shallow(<UserProfile />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should save a snapshot of the component', () => {
    expect(component).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const { onPress, email } = component.find(UserEmail).props();
    const image = component.find(UserAvatar).prop('image');
    if (onPress) {
      onPress(email);
      expect(callback).toHaveBeenCalledTimes(1);
    }
    expect(image).toBe(sessionMock.user?.avatar);
  });
});
