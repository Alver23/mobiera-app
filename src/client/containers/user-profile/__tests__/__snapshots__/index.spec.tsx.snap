// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfile /> should save a snapshot of the component 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
    }
  }
>
  <Themed.Card
    containerStyle={
      Array [
        Object {
          "borderWidth": 0,
          "margin": 0,
          "padding": 0,
        },
        Object {
          "backgroundColor": "#242424",
        },
      ]
    }
  >
    <UserAvatar
      image="http://localhost/img.png"
    />
  </Themed.Card>
  <Themed.Card
    containerStyle={
      Array [
        Object {
          "borderWidth": 0,
          "margin": 0,
          "padding": 0,
        },
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <Themed.Text
      h4={true}
      style={
        Object {
          "color": "#242424",
          "marginVertical": 12,
        }
      }
    >
      fake name
    </Themed.Text>
    <UserEmail
      email="fake@fake.com"
      onPress={[Function]}
    />
  </Themed.Card>
</View>
`;
